### DISCUSSION TITLE
Some questions about parking env (HER + SAC)

### DISCUSSION THREAD

**Discussion by FieClaussell123** [eleurent/highway-env#164] [General]
Hi, i have some questions when i implement ```parking env```.

I find that the vehicle sometimes reverses for reaching the goal, and i  want to know when it will act like that. But I cannot fully understand the code. May i ask you some questions ?

- Does the vehicle consider the position of itself and the goal (or reward) to decide what action to take?   or just take random action  and remember what action can get more reward, repeat the action next time at that position. 

I thought the vehicle will take the action which make the reward larger (minimize the distance between the vehicle and the goal) every steps. But sometime it will reverse, make the reward smaller. I realize that my idea is wrong.

**Comment by eleurent** [comment]
To decide what action to take, the vehicle gets to observe its current state (position, velocity and heading) and its goal state (target position, zero velocity, target heading). But to learn how to reach a goal, it must rely on the history of previous actions (initially random) and rewards (that depend on the goal) that it has seen so far.

A reason for the vehicle to reverse is that the goals at specified in terms of both location and heading. Since the vehicle starts with a random heading, it may be facing backward to its goal location. In that case, it is possible that it starts driving in reverse before turning around.
When doing so, the immediate rewards will temporarily decrease because the vehicle gets further from its goal by maneuvering, but it is worth it in the long run in terms of cumulative reward (aka return / value), since the vehicle can then reach the target location AND heading, yielding the maximum possible rewards.

I hope that was clear

**Comment by FieClaussell123** [comment]
Thanks for your answers !  But there are some questions I still can't understand.

I want to know the part of the cumulative reward, where can i find the descriptions in the code ?

And I found that the closer the position of goal to the initial position of the vehicle, the more likely it is to fail. I think the reason for this is because it does not have enough space to turn around.

Do you have some ideas to solve the problem ?

**Comment by eleurent** [comment]
In Reinforcement Learning, you optimize the (discounted) sum of rewards. For parking-v0, the reward function is defined here: https://github.com/eleurent/highway-env/blob/master/highway_env/envs/parking_env.py#L90

I haven't observed myself that it fails when the goal is close to the initial position of the vehicle, so I'm not sure how to help.

**Comment by FieClaussell123** [comment]
oh .. maybe i did something wrong. Is it possible that I always get 100% success rate on this env ?

**Comment by eleurent** [comment]
Yes it is possible. You can try running [this colab notebook](https://colab.research.google.com/github/eleurent/highway-env/blob/master/scripts/parking_her.ipynb) for example, it should reach >95% success rate.
